### ISSUE TITLE
Swift代码库进行lora checkpoint的continue sft，加载模型和checkpoint后可训练参数为0%

### ISSUE THREAD

**Issue by lilqz66** [modelscope/ms-swift#4723]
**Describe the bug**
What the bug is, and how to reproduce, better with screenshots(描述bug以及复现过程，最好有截图)

**Your hardware and system info**
Write your system info like CUDA version/system/GPU/torch version here(在这里给出硬件信息和系统信息，如CUDA版本，系统，GPU型号和torch版本等)

**Additional context**
Add any other context about the problem here(在这里补充其他信息)

**Comment by lilqz66** [comment]
代码：

import os
import json
import argparse
from functools import partial
from dataclasses import dataclass, field, asdict

from swift.llm import get_model_tokenizer, load_dataset, get_template, EncodePreprocessor
from swift.utils import get_logger, find_all_linears, get_model_parameter_info, plot_images, seed_everything
from swift.tuners import Swift, LoraConfig
from swift.trainers import Seq2SeqTrainer, Seq2SeqTrainingArguments


from cscd_origin_dataset import CscdOriginPreprocessor
# from cscd_dataset import CscdPreprocessor
from swift.llm import DatasetMeta, load_dataset, register_dataset

os.environ['CUDA_DEVICE_ORDER'] = 'PCI_BUS_ID'


@dataclass
class Config:
    last_model_checkpoint: str = None
    model_path: str = ''
    output_dir: str = ''
    dataset_path: str = ''
    datasets: list = field(default_factory=lambda: ['CSCD-NS/train'])
    system_prompt: str = 'You are a helpful assistant.'
    max_length: int = 2048
    data_seed: int = 42
    split_ratio: float = 0.1
    num_proc: int = 4
    lora_rank: int = 8
    lora_alpha: int = 32
    cuda_devices: str = '2'

    batch_size: int = 8
    gradient_checkpointing: bool = True
    learning_rate: float = 1e-4
    weight_decay: float = 0.1
    lr_scheduler: str = 'cosine'
    warmup_ratio: float = 0.05
    epochs: int = 2
    save_strategy: str = 'epoch'
    save_steps: int = 100
    eval_strategy: str = 'epoch'
    eval_steps: int = 100
    grad_accum: int = 4
    log_steps: int = 5
    save_limit: int = 6
    dataloader_workers: int = 1
    report_to: list = field(default_factory=lambda: ['tensorboard'])


class TrainerPipeline:
    def __init__(self, cfg: Config):
        self.cfg = cfg
        os.environ['CUDA_VISIBLE_DEVICES'] = cfg.cuda_devices
        self.logger = get_logger()
        seed_everything(cfg.data_seed)

    def load_model(self):
        self.logger.info('Loading model and tokenizer...')
        # 加载原始模型
        model, tokenizer = get_model_tokenizer(self.cfg.model_path)

        # LoRA 配置
        target_modules = find_all_linears(model)
        lora_config = LoraConfig(
            task_type='CAUSAL_LM',
            r=self.cfg.lora_rank,
            lora_alpha=self.cfg.lora_alpha,
            target_modules=target_modules
        )

        # 注入 LoRA adapter（恢复训练）
        if self.cfg.last_model_checkpoint:
            model = Swift.from_pretrained(
                model=model,
                model_id=self.cfg.last_model_checkpoint,
                lora_config=lora_config
            )
        else:
            model = Swift.prepare_model(model, lora_config)

        template = get_template(
            model.model_meta.template, tokenizer,
            default_system=self.cfg.system_prompt,
            max_length=self.cfg.max_length
        )
        template.set_mode('train')

        info = get_model_parameter_info(model)
        self.logger.info(f'Model parameters: {info}')

        self.model, self.tokenizer, self.template = model, tokenizer, template
        trainable_params = [n for n, p in self.model.named_parameters() if p.requires_grad]
        print(f"Trainable parameters: {trainable_params}")

    def prepare_data(self):
        
        register_dataset(
            DatasetMeta(
                dataset_path=self.cfg.dataset_path,
                dataset_name=self.cfg.datasets[0],
                preprocess_func=CscdOriginPreprocessor(),
            ))

        self.logger.info('Loading and preprocessing dataset...')
        train_ds, val_ds = load_dataset(
            self.cfg.datasets,
            split_dataset_ratio=self.cfg.split_ratio,
            num_proc=self.cfg.num_proc,
            seed=self.cfg.data_seed
        )
        encoder = EncodePreprocessor(template=self.template)
        self.train_dataset = encoder(train_ds, num_proc=self.cfg.num_proc)
        self.val_dataset = encoder(val_ds, num_proc=self.cfg.num_proc)

    def setup_trainer(self):
        self.logger.info('Configuring training arguments...')
        args = Seq2SeqTrainingArguments(
            output_dir=self.cfg.output_dir,
            learning_rate=self.cfg.learning_rate,
            per_device_train_batch_size=self.cfg.batch_size,
            per_device_eval_batch_size=self.cfg.batch_size,
            gradient_checkpointing=self.cfg.gradient_checkpointing,
            weight_decay=self.cfg.weight_decay,
            lr_scheduler_type=self.cfg.lr_scheduler,
            warmup_ratio=self.cfg.warmup_ratio,
            report_to=self.cfg.report_to,
            logging_first_step=True,
            save_strategy=self.cfg.save_strategy,
            save_steps=self.cfg.save_steps,
            eval_strategy=self.cfg.eval_strategy,
            eval_steps=self.cfg.eval_steps,
            gradient_accumulation_steps=self.cfg.grad_accum,
            num_train_epochs=self.cfg.epochs,
            metric_for_best_model='loss',
            save_total_limit=self.cfg.save_limit,
            logging_steps=self.cfg.log_steps,
            dataloader_num_workers=self.cfg.dataloader_workers,
        )

        self.logger.info('Initializing trainer...')
        self.model.enable_input_require_grads()
        self.trainer = Seq2SeqTrainer(
            model=self.model,
            args=args,
            data_collator=self.template.data_collator,
            train_dataset=self.train_dataset,
            eval_dataset=self.val_dataset,
            template=self.template,
        )

    def train(self):
        self.trainer.train()
        ckpt = self.trainer.state.last_model_checkpoint
        self.logger.info(f'Last checkpoint: {ckpt}')
        images_dir = os.path.join(self.cfg.output_dir, 'images')
        plot_images(images_dir, self.trainer.args.logging_dir, ['train/loss'], 0.9)


def str2bool(v):
    if isinstance(v, bool): return v
    return v.lower() in ('true', '1', 'yes')


import dataclasses

def parse_args():
    parser = argparse.ArgumentParser()
    for field_ in Config.__dataclass_fields__.values():
        arg_type = field_.type
        if field_.default_factory is not dataclasses.MISSING:
            default = field_.default_factory()
        else:
            default = field_.default
        if arg_type == bool:
            parser.add_argument(f'--{field_.name}', type=str2bool, default=default)
        elif arg_type == list:
            parser.add_argument(f'--{field_.name}', type=lambda s: s.split(','), default=default)
        else:
            parser.add_argument(f'--{field_.name}', type=arg_type, default=default)

    args = parser.parse_args()
    return Config(**vars(args))



def main():
    cfg = parse_args()

    # 打印所有配置参数
    print('\n========== Training Configuration ==========')
    for k, v in asdict(cfg).items():
        print(f'{k:30s}: {v}')
    print('============================================\n')

    # 保存配置到指定 JSON 文件
    os.makedirs(cfg.output_dir, exist_ok=True)
    config_path = os.path.join(cfg.output_dir, 'train_args_print.json')
    with open(config_path, 'w', encoding='utf-8') as f:
        json.dump(asdict(cfg), f, ensure_ascii=False, indent=2)
    print(f'Config saved to: {config_path}\n')

    pipeline = TrainerPipeline(cfg)
    pipeline.load_model()
    pipeline.prepare_data()
    pipeline.setup_trainer()
    pipeline.train()


if __name__ == '__main__':
    main()


打印信息：
[INFO:swift] default_system: 'You are a helpful assistant.'
[INFO:swift] max_length: 2048
[INFO:swift] response_prefix: ''
[INFO:swift] agent_template: hermes
[INFO:swift] norm_bbox: norm1000
[INFO:swift] Model parameters: PeftModelForCausalLM: 1729.2913M Params (0.0000M Trainable [0.0000%]), 0.0001M Buffers.
Trainable parameters: []


查看文档没有使用swift库进行lora continue sft的样例，可以帮忙解决问题或者给出lora continue sft的样例吗

**Comment by lilqz66** [comment]
代码：
`import os
import json
import argparse
from functools import partial
from dataclasses import dataclass, field, asdict

from swift.llm import get_model_tokenizer, load_dataset, get_template, EncodePreprocessor
from swift.utils import get_logger, find_all_linears, get_model_parameter_info, plot_images, seed_everything
from swift.tuners import Swift, LoraConfig
from swift.trainers import Seq2SeqTrainer, Seq2SeqTrainingArguments


from cscd_origin_dataset import CscdOriginPreprocessor
# from cscd_dataset import CscdPreprocessor
from swift.llm import DatasetMeta, load_dataset, register_dataset

os.environ['CUDA_DEVICE_ORDER'] = 'PCI_BUS_ID'


@dataclass
class Config:
    last_model_checkpoint: str = None
    model_path: str = ''
    output_dir: str = ''
    dataset_path: str = ''
    datasets: list = field(default_factory=lambda: ['CSCD-NS/train'])
    system_prompt: str = 'You are a helpful assistant.'
    max_length: int = 2048
    data_seed: int = 42
    split_ratio: float = 0.1
    num_proc: int = 4
    lora_rank: int = 8
    lora_alpha: int = 32
    cuda_devices: str = '2'

    batch_size: int = 8
    gradient_checkpointing: bool = True
    learning_rate: float = 1e-4
    weight_decay: float = 0.1
    lr_scheduler: str = 'cosine'
    warmup_ratio: float = 0.05
    epochs: int = 2
    save_strategy: str = 'epoch'
    save_steps: int = 100
    eval_strategy: str = 'epoch'
    eval_steps: int = 100
    grad_accum: int = 4
    log_steps: int = 5
    save_limit: int = 6
    dataloader_workers: int = 1
    report_to: list = field(default_factory=lambda: ['tensorboard'])


class TrainerPipeline:
    def __init__(self, cfg: Config):
        self.cfg = cfg
        os.environ['CUDA_VISIBLE_DEVICES'] = cfg.cuda_devices
        self.logger = get_logger()
        seed_everything(cfg.data_seed)

    def load_model(self):
        self.logger.info('Loading model and tokenizer...')
        # 加载原始模型
        model, tokenizer = get_model_tokenizer(self.cfg.model_path)

        # LoRA 配置
        target_modules = find_all_linears(model)
        lora_config = LoraConfig(
            task_type='CAUSAL_LM',
            r=self.cfg.lora_rank,
            lora_alpha=self.cfg.lora_alpha,
            target_modules=target_modules
        )

        # 注入 LoRA adapter（恢复训练）
        if self.cfg.last_model_checkpoint:
            model = Swift.from_pretrained(
                model=model,
                model_id=self.cfg.last_model_checkpoint,
                lora_config=lora_config
            )
        else:
            model = Swift.prepare_model(model, lora_config)

        template = get_template(
            model.model_meta.template, tokenizer,
            default_system=self.cfg.system_prompt,
            max_length=self.cfg.max_length
        )
        template.set_mode('train')

        info = get_model_parameter_info(model)
        self.logger.info(f'Model parameters: {info}')

        self.model, self.tokenizer, self.template = model, tokenizer, template
        trainable_params = [n for n, p in self.model.named_parameters() if p.requires_grad]
        print(f"Trainable parameters: {trainable_params}")

    def prepare_data(self):
        
        register_dataset(
            DatasetMeta(
                dataset_path=self.cfg.dataset_path,
                dataset_name=self.cfg.datasets[0],
                preprocess_func=CscdOriginPreprocessor(),
            ))

        self.logger.info('Loading and preprocessing dataset...')
        train_ds, val_ds = load_dataset(
            self.cfg.datasets,
            split_dataset_ratio=self.cfg.split_ratio,
            num_proc=self.cfg.num_proc,
            seed=self.cfg.data_seed
        )
        encoder = EncodePreprocessor(template=self.template)
        self.train_dataset = encoder(train_ds, num_proc=self.cfg.num_proc)
        self.val_dataset = encoder(val_ds, num_proc=self.cfg.num_proc)

    def setup_trainer(self):
        self.logger.info('Configuring training arguments...')
        args = Seq2SeqTrainingArguments(
            output_dir=self.cfg.output_dir,
            learning_rate=self.cfg.learning_rate,
            per_device_train_batch_size=self.cfg.batch_size,
            per_device_eval_batch_size=self.cfg.batch_size,
            gradient_checkpointing=self.cfg.gradient_checkpointing,
            weight_decay=self.cfg.weight_decay,
            lr_scheduler_type=self.cfg.lr_scheduler,
            warmup_ratio=self.cfg.warmup_ratio,
            report_to=self.cfg.report_to,
            logging_first_step=True,
            save_strategy=self.cfg.save_strategy,
            save_steps=self.cfg.save_steps,
            eval_strategy=self.cfg.eval_strategy,
            eval_steps=self.cfg.eval_steps,
            gradient_accumulation_steps=self.cfg.grad_accum,
            num_train_epochs=self.cfg.epochs,
            metric_for_best_model='loss',
            save_total_limit=self.cfg.save_limit,
            logging_steps=self.cfg.log_steps,
            dataloader_num_workers=self.cfg.dataloader_workers,
        )

        self.logger.info('Initializing trainer...')
        self.model.enable_input_require_grads()
        self.trainer = Seq2SeqTrainer(
            model=self.model,
            args=args,
            data_collator=self.template.data_collator,
            train_dataset=self.train_dataset,
            eval_dataset=self.val_dataset,
            template=self.template,
        )

    def train(self):
        self.trainer.train()
        ckpt = self.trainer.state.last_model_checkpoint
        self.logger.info(f'Last checkpoint: {ckpt}')
        images_dir = os.path.join(self.cfg.output_dir, 'images')
        plot_images(images_dir, self.trainer.args.logging_dir, ['train/loss'], 0.9)


def str2bool(v):
    if isinstance(v, bool): return v
    return v.lower() in ('true', '1', 'yes')


import dataclasses

def parse_args():
    parser = argparse.ArgumentParser()
    for field_ in Config.__dataclass_fields__.values():
        arg_type = field_.type
        if field_.default_factory is not dataclasses.MISSING:
            default = field_.default_factory()
        else:
            default = field_.default
        if arg_type == bool:
            parser.add_argument(f'--{field_.name}', type=str2bool, default=default)
        elif arg_type == list:
            parser.add_argument(f'--{field_.name}', type=lambda s: s.split(','), default=default)
        else:
            parser.add_argument(f'--{field_.name}', type=arg_type, default=default)

    args = parser.parse_args()
    return Config(**vars(args))



def main():
    cfg = parse_args()

    # 打印所有配置参数
    print('\n========== Training Configuration ==========')
    for k, v in asdict(cfg).items():
        print(f'{k:30s}: {v}')
    print('============================================\n')

    # 保存配置到指定 JSON 文件
    os.makedirs(cfg.output_dir, exist_ok=True)
    config_path = os.path.join(cfg.output_dir, 'train_args_print.json')
    with open(config_path, 'w', encoding='utf-8') as f:
        json.dump(asdict(cfg), f, ensure_ascii=False, indent=2)
    print(f'Config saved to: {config_path}\n')

    pipeline = TrainerPipeline(cfg)
    pipeline.load_model()
    pipeline.prepare_data()
    pipeline.setup_trainer()
    pipeline.train()


if __name__ == '__main__':
    main()
`

打印信息：
[INFO:swift] default_system: 'You are a helpful assistant.'
[INFO:swift] max_length: 2048
[INFO:swift] response_prefix: ''
[INFO:swift] agent_template: hermes
[INFO:swift] norm_bbox: norm1000
[INFO:swift] Model parameters: PeftModelForCausalLM: 1729.2913M Params (0.0000M Trainable [0.0000%]), 0.0001M Buffers.
Trainable parameters: []

查看文档没有使用swift库进行lora continue sft的样例，可以帮忙解决问题或者给出lora continue sft的样例吗

**Comment by lilqz66** [comment]
代码：
    def load_model(self):
        self.logger.info('Loading model and tokenizer...')
        # 加载原始模型
        model, tokenizer = get_model_tokenizer(self.cfg.model_path)

        # LoRA 配置
        target_modules = find_all_linears(model)
        lora_config = LoraConfig(
            task_type='CAUSAL_LM',
            r=self.cfg.lora_rank,
            lora_alpha=self.cfg.lora_alpha,
            target_modules=target_modules
        )

        # 注入 LoRA adapter（恢复训练）
        if self.cfg.last_model_checkpoint:
            model = Swift.from_pretrained(
                model=model,
                model_id=self.cfg.last_model_checkpoint,
                lora_config=lora_config
            )
        else:
            model = Swift.prepare_model(model, lora_config)

        template = get_template(
            model.model_meta.template, tokenizer,
            default_system=self.cfg.system_prompt,
            max_length=self.cfg.max_length
        )
        template.set_mode('train')

        info = get_model_parameter_info(model)
        self.logger.info(f'Model parameters: {info}')

        self.model, self.tokenizer, self.template = model, tokenizer, template
        trainable_params = [n for n, p in self.model.named_parameters() if p.requires_grad]
        print(f"Trainable parameters: {trainable_params}")



打印信息：
[INFO:swift] default_system: 'You are a helpful assistant.'
[INFO:swift] max_length: 2048
[INFO:swift] response_prefix: ''
[INFO:swift] agent_template: hermes
[INFO:swift] norm_bbox: norm1000
[INFO:swift] Model parameters: PeftModelForCausalLM: 1729.2913M Params (0.0000M Trainable [0.0000%]), 0.0001M Buffers.
Trainable parameters: []

查看文档没有使用swift库进行lora continue sft的样例，可以帮忙解决问题或者给出lora continue sft的样例吗

**Comment by Jintao-Huang** [comment]
请使用 命令行训练


你也可以通过 sft_main, TrainArguments 更上层分装进行训练
